openssh/key-regen.sh

```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+# set -ex
+
+if [ "x${EUID}" != "x0" ]; then
+    echo "Must be root! Aborting..."
+    exit 1
+fi
+
+# Get all active identity files from ssh_config and generate a file for each
+ID_DIR="/home/${OQS_USER}/.ssh"
+readarray ID_ALGS <<< $(sed -n "s:^identityfile.*/id_::Ip" ${OQS_INSTALL_DIR}/ssh_config)
+# Find longest name
+MAX_LEN=0
+for alg in ${ID_ALGS[@]}; do
+    if [ ${#alg} -gt $MAX_LEN ]; then
+        MAX_LEN=${#alg}
+    fi
+done
+echo "Generating identity files as configured in ${OQS_INSTALL_DIR}/ssh_config:"
+for alg in ${ID_ALGS[@]}; do
+    printf "\t%-$((MAX_LEN + 1))s" ${alg^^}
+    if [ $alg == "rsa" ] || [ $alg == "dsa" ] || [ $alg == "ecdsa*" ] || [ $alg == "ed25519*" ]; then
+        alg_pre=''
+    else
+        alg_pre='ssh-'
+    fi
+    ID_FILE=${ID_DIR}/id_${alg}
+    if [ -e $ID_FILE ]; then
+        printf "exists @ %-$((MAX_LEN + ${#ID_DIR} + 10))s --> SKIPPED\n" "$ID_FILE(.pub)"
+    else
+        CMD="su ${OQS_USER} -c \"${OQS_INSTALL_DIR}/bin/ssh-keygen -t $alg_pre$(echo $alg | sed 's/_/-/g') -f $ID_FILE -N '' -q\""
+        # echo $CMD
+        eval $CMD
+        if [ $? -ne 0 ]; then
+            echo "FAILED"
+        else
+            echo "generated @ $ID_FILE(.pub)"
+        fi
+    fi
+done
+# echo " done!"
+
+# Regenerate existing host keys
+HOST_KEY_DIR=$(echo $OQS_INSTALL_DIR | sed 's:/*$::')
+echo -e "\nGenerating host key files as configured in $OQS_INSTALL_DIR/sshd_config (all HostKeyAlgorithms):"
+
+# Get algorithms from sshd_config
+IFS=',' read -ra HOST_KEY_ALGS <<< $(sed -n "s/^hostkeyalgorithms[ \t=]*//Ip" ${OQS_INSTALL_DIR}/sshd_config)
```

Comment:
Do I understand this correct in that one has to manually update the set of supported algorithms in sshd_config as and when the algorithm list supported upstream changes? Might it be worth while thinking about some automation for this? But then again, it's not a regular occurrence...

Reply:
You do understand this correctly.

The reason I ended up with this version is that I didn't automatically want to generate all host keys with `ssh-keygen -A`  (which would not solve the problem with the config files) because it takes quite some time especially with the rainbow keys. For the same reason I use the build target `make install-nokeys` when compiling ssh while building the image, it speeds up building considerably. So I needed some place to configure the algorithms that are used per default, because this affects the generation of the keys as well as the configuration of sshd and ssh.

One possibility would be adding a feature to specify the algorithms at build time and automatically update sshd_config and ssh_config accordingly in the Dockerfile, using sed or something similar. Thinking about it, I will have a short try with this.

And I just realized that I diverged from what you were actually asking: For now I don't see another possibility to automate this without more complex scripting that checks upstream algorithms changes and applies them. But as you say, it's quite irregular and as we have already progressed quite far in the selection of PQC algorithms, there probably won't be that many changes before the new standards are released.

Reply:
Agreed. But maybe a footnote stating this would be good. Or maybe even better a test that would/will break if the upstream algorithm list changes.

Reply:
If you are aiming to only update this for major releases of oqs/openssh-oqs (e.g., v0.4.0 in this case), than the manual process might be ok. Agreed with @baentsch that a failsafe test would be good.

Reply:
I added the option to change the install target at build time, and parametrized the liboqs and oqs-openssh version in the Dockerfile, so it is clearer what version is being used.

> Agreed. But maybe a footnote stating this would be good. Or maybe even better a test that would/will break if the upstream algorithm list changes.

I will mention this in the Readme, and also in form of a comment in the Dockerfile.  As I explicitely check out the 0.4.0 release, nothing would change without a user actually making a change AFAIK. So I suppose mentioning it should suffice. If you agree I will leave it at this, also because I don't have an angle on how to automatically test this, would this be a part of the CCI config? It could notify that there is a new release of liboqs, so it wouldn't slip through.